hadoop-hdds/client/src/main/java/org/apache/hadoop/hdds/scm/storage/StreamBlockInput.java

```diff
@@ -0,0 +1,777 @@
+
+/**
+ * An {@link java.io.InputStream} called from KeyInputStream to read a block from the
+ * container.
+ */
+public class StreamBlockInput extends BlockExtendedInputStream
+    implements Seekable, CanUnbuffer, ByteBufferReadable {
+  private static final Logger LOG =
+      LoggerFactory.getLogger(StreamBlockInput.class);
+  private final BlockID blockID;
+  private final long length;
+  private final AtomicReference<Pipeline> pipelineRef =
+      new AtomicReference<>();
+  private final AtomicReference<Token<OzoneBlockTokenIdentifier>> tokenRef =
+      new AtomicReference<>();
+  private XceiverClientFactory xceiverClientFactory;
+  private XceiverClientSpi xceiverClient;
+
+  private List<Long> bufferoffsets;
+  private int bufferIndex;
+  private long blockPosition = -1;
+  private List<ByteBuffer> buffers;
+  private boolean allocated = false;
+  private long bufferOffsetWrtBlockDataData;
+  private long buffersSize;
+  private static final int EOF = -1;
+  private final List<XceiverClientSpi.Validator> validators;
+  private final boolean verifyChecksum;
+  private final Function<BlockID, BlockLocationInfo> refreshFunction;
+  private final RetryPolicy retryPolicy;
+  private int retries;
+
+
+  public StreamBlockInput(
+      BlockID blockID, long length, Pipeline pipeline,
+      Token<OzoneBlockTokenIdentifier> token,
+      XceiverClientFactory xceiverClientFactory,
+      Function<BlockID, BlockLocationInfo> refreshFunction,
+      OzoneClientConfig config) throws IOException {
+    this.blockID = blockID;
+    LOG.debug("Initializing StreamBlockInput for block {}", blockID);
+    this.length = length;
+    setPipeline(pipeline);
+    tokenRef.set(token);
+    this.xceiverClientFactory = xceiverClientFactory;
+    this.validators = ContainerProtocolCalls.toValidatorList(
+        (request, response) -> validateBlock(response));
+    this.verifyChecksum = config.isChecksumVerify();
+    this.refreshFunction = refreshFunction;
+    this.retryPolicy =
+        HddsClientUtils.createRetryPolicy(config.getMaxReadRetryCount(),
+            TimeUnit.SECONDS.toMillis(config.getReadRetryInterval()));
+
+  }
+
+
+  public BlockID getBlockID() {
+    return blockID;
+  }
+
+  public long getLength() {
+    return length;
+  }
+
+  @Override
+  public synchronized long getPos() {
+    if (length == 0) {
+      return 0;
+    }
+    if (blockPosition >= 0) {
+      return blockPosition;
+    }
+
+    if (allocated && !buffersHaveData() && !dataRemainingInBlock()) {
+      Preconditions.checkState(
+          bufferOffsetWrtBlockDataData + buffersSize == length,
+          "EOF detected but not at the last byte of the chunk");
+      return length;
+    }
+    if (buffersHaveData()) {
+      // BufferOffset w.r.t to BlockData + BufferOffset w.r.t buffers +
+      // Position of current Buffer
+      return bufferOffsetWrtBlockDataData + bufferoffsets.get(bufferIndex) +
+          buffers.get(bufferIndex).position();
+    }
```

Comment:
We can switch these blocks and simplify the if statement.
```suggestion
    if (buffersHaveData()) {
      // BufferOffset w.r.t to BlockData + BufferOffset w.r.t buffers +
      // Position of current Buffer
      return bufferOffsetWrtBlockDataData + bufferoffsets.get(bufferIndex) +
          buffers.get(bufferIndex).position();
    }
    if (allocated && !dataRemainingInBlock()) {
      Preconditions.checkState(
          bufferOffsetWrtBlockDataData + buffersSize == length,
          "EOF detected but not at the last byte of the chunk");
      return length;
    }
```

This was we already know if the `allocated && !dataRemainingInBlock()` we do not have buffer data as otherwise it would have matched buffershaveData() condition before